src/threading.mjs

```diff
@@ -0,0 +1,92 @@
+import {
+  Worker,
+  isMainThread,
+  parentPort,
+  workerData,
+} from 'node:worker_threads';
+import { allGenerators } from './generators/index.mjs';
+
+// If inside a worker thread, perform the generator logic here
+if (!isMainThread) {
+  const { name, dependencyOutput, extra } = workerData;
+  const generator = allGenerators[name];
+
+  // Execute the generator and send the result or error back to the parent thread
+  generator
+    .generate(dependencyOutput, extra)
+    .then(result => parentPort.postMessage(result))
+    .catch(error => parentPort.postMessage({ error }));
+}
+
+/**
+ * WorkerPool class to manage a pool of worker threads
+ */
+export class WorkerPool {
+  /** @private {number} - Number of active threads */
+  activeThreads = 0;
+  /** @private {Array<Function>} - Queue of pending tasks */
+  queue = [];
+
+  /**
+   * Runs a generator within a worker thread.
+   * @param {string} name - The name of the generator to execute
+   * @param {any} dependencyOutput - Input data for the generator
+   * @param {number} threads - Maximum number of threads to run concurrently
+   * @param {Object} extra - Additional options for the generator
+   * @returns {Promise<any>} Resolves with the generator result, or rejects with an error
+   */
+  run(name, dependencyOutput, threads, extra) {
+    return new Promise((resolve, reject) => {
+      /**
+       * Function to run the generator in a worker thread
+       */
+      const run = () => {
+        this.activeThreads++;
```

Comment:
Can we use atomics, please?

Reply:
What for? Only the main thread is updating `activeThreads`

Reply:
Still, it is good to be an atomic, This value also can get decremented, and we definitely want to ensure no race conditions happen.